Update cert-manager dependency to v1.4.0 (#16)
projectsyn/component-cert-manager: docs/modules/ROOT/pages/how-tos/upgrade-v1-v2.adoc

= Upgrade `component-cert-manager` from `v1.0.0` to `v2.0.0`

> This upgrade upgrades the underlying `cert-manager` helm chart from `v0.15.1` to `v1.4.0`. There are some breaking changes in `cert-manager`. More information can be found https://github.com/jetstack/cert-manager/releases[here] and https://cert-manager.io/docs/installation/upgrading/[here].

Reviewer: Please follow the style guide https://syn.tools/syn/references/style-guide.html:
1 line per sentence
```suggestion
This upgrade upgrades the underlying `cert-manager` helm chart from `v0.x` to `v1.x`. 
There are some breaking changes in `cert-manager`. 
More information can be found https://github.com/jetstack/cert-manager/releases[in changelog on GitHub] and in the https://cert-manager.io/docs/installation/upgrading/[upgrade documentation].
```

Author: ok thanks!

> This upgrade upgrades the underlying `cert-manager` helm chart from `v0.15.1` to `v1.4.0`. There are some breaking changes in `cert-manager`. More information can be found https://github.com/jetstack/cert-manager/releases[here] and https://cert-manager.io/docs/installation/upgrading/[here].

Author: totally agree :)

The upgrade is done with the following steps:

. Prepare CRD upgrade
. Compile Commodore catalog with `component-cert-manager` v2.x
. Verify cert-manager upgrade

== Prerequisites

. `commodore`
. `kubectl` or `oc`
. https://github.com/mikefarah/yq[`yq`] version 4 (alternatively, any editor works)

> == Prepare CRD upgrade

Reviewer: I think we can all keep it in one section and use numeric steps instead.
```suggestion
== Upgrade
```

> This step updates the `"kubectl.kubernetes.io/last-applied-configuration"` annotation which is required for a successful 3-way merge.
> 
> [source,bash]

Reviewer: ```suggestion
. Prepare CRD upgrade
+
[source,bash]
```

----
cat <<'EOF' > cert-manager-prepare-merge-crds.sh
#!/bin/sh

set -euo pipefail

for crd in certificaterequests.cert-manager.io certificates.cert-manager.io challenges.acme.cert-manager.io clusterissuers.cert-manager.io issuers.cert-manager.io orders.acme.cert-manager.io; do
  # Get CRD definition in YAML
  kubectl get crd "${crd}" -o yaml > "${crd}.yaml"
  # Remove all metadata properties except `metadata.name`
  yq -i eval 'del(.status) | del(.metadata) | .metadata.name += "'${crd}'"' "${crd}.yaml"

>   # Apply the CRD again (this shouldn't change anything, except adding the annotation "kubectl.kubernetes.io/last-applied-configuration")

Reviewer: ```suggestion
  # Apply the CRD again (this shouldn't change anything, except updating the annotation "kubectl.kubernetes.io/last-applied-configuration")
```

  # You will also see some warnings in the output mentioning the annotation.
  # This is expected and actually required.
  kubectl apply -f "${crd}.yaml"
done
EOF

sh cert-manager-prepare-merge-crds.sh
----

> 
> == Compile Commodore catalog with `component-cert-manager` v2.x
> 
> . Compile and push cluster catalog

Reviewer: ```suggestion

. Compile and push cluster catalog with `component-cert-manager` v2.x
```

== Verify cert-manager and certificates

> . Verify successful sync and health status in ArgoCD

Reviewer: I think it's worth adding a note that force sync shouldn't be done.
```suggestion
. Verify successful sync and health status in ArgoCD
+
[WARNING]
====
Under no circumstances should you trigger a force sync in ArgoCD.
This causes ArgoCD to remove and recreate the CRDs, in which case all attached resources get deleted.
====
```

> = Upgrade `component-cert-manager` from `v1.x` to `v2.x`

Reviewer: The how-to page needs to be linked in the `partial/nav.adoc` file in order to pop up in the navigation

Author: done, thanks!

> Version 2.x upgrades the underlying `cert-manager` helm chart from `v0.x` to `v1.x`.

Reviewer 2: ```suggestion
Version 2.x upgrades the underlying `cert-manager` Helm chart from `v0.x` to `v1.x`.
```

> This upgrade upgrades the underlying `cert-manager` helm chart from `v0.x` to `v1.x`.

Reviewer 3: ```suggestion
Version 2.0.0 upgrades the underlying `cert-manager` helm chart from `v0.x` to `v1.x`.
```

"upgrade upgrades" reads bumbpy.

Author: agreed 👍

There are some breaking changes in `cert-manager`.
More information can be found https://github.com/jetstack/cert-manager/releases[in changelog on GitHub] and in the https://cert-manager.io/docs/installation/upgrading/[upgrade documentation].

The upgrade is done with the following steps:

. Compile Commodore catalog with `component-cert-manager` v2.x
. Fix CRD upgrade
. Verify cert-manager upgrade

== Prerequisites

. `commodore`
. `kubectl` or `oc`
. https://github.com/mikefarah/yq[`yq`] version 4 (alternatively, any editor works)

== Upgrade

. Compile and push cluster catalog with `component-cert-manager` v2.x

> . Fix CRD upgrade

Reviewer 2: Is that something which could be automated somehow?

Reviewer: I'm not sure. We could hack something with [ArgoCD hooks](https://argoproj.github.io/argo-cd/user-guide/resource_hooks/), but it comes with its own pitfalls. We'd have to ensure that the hook is idempotent and compatible for any upgrades within v2.x versions. Even the ServiceAccount executing the Job would need to be granted cluster admin permissions in order to modify CRDs.

Author: We could try with hooks. The hooks would need to be of type `PreSync` or in an earlier wave for Argo not to fail before executing the hooks.

The fix should be idempotent and not too hard to keep compatible since it does essentially nothing 😅.

Reviewer: Yes, but maybe you need another half-day to a day to engineer a working hook with the cluster-admin permission and docker image it requires to perform the hook

Reviewer 2: If you have a component like this which might be deployed to a huge amount of clusters I don't think it's feasible if you have to manually touch every cluster to upgrade the component. I would find it a valuable time investment to try to automate the upgrade.

Author: Started automating the process in #18.

+
[source,bash]
----
cat <<'EOF' > cert-manager-prepare-merge-crds.sh
#!/bin/sh

set -euo pipefail

for crd in certificaterequests.cert-manager.io certificates.cert-manager.io challenges.acme.cert-manager.io clusterissuers.cert-manager.io issuers.cert-manager.io orders.acme.cert-manager.io; do
  # Get CRD definition in YAML
  kubectl get crd "${crd}" -o yaml > "${crd}.yaml"
  # Remove all metadata properties except `metadata.name`
  yq -i eval 'del(.status) | del(.metadata) | .metadata.name += "'${crd}'"' "${crd}.yaml"
  # Apply the CRD again (this shouldn't change anything, except updating the annotation "kubectl.kubernetes.io/last-applied-configuration")
  # You will also see some warnings in the output mentioning the annotation.
  # This is expected and actually required.
  kubectl apply -f "${crd}.yaml"
done
EOF

sh cert-manager-prepare-merge-crds.sh
----

== Verify cert-manager and certificates

. Verify successful sync and health status in ArgoCD
+
[NOTE]
====
If ArgoCD sync is marked as failed it may be required to trigger another sync.
====
+

> [WARNING]

Reviewer 3: ```suggestion
[IMPORTANT]
```

This causes ArgoCD to remove and recreate the CRDs, in which case all attached resources get deleted.
====
. Check for obvious deployment errors
+
[source,bash]
----
kubectl -n syn-cert-manager get pods
kubectl -n syn-cert-manager logs -l "app.kubernetes.io/name=cert-manager"
----

. Optional: Create testing ingress with cert-manager managed certificate.
+

> [NOTE]
> ====
> The cluster has to be publicly reachable and must have a wildcard DNS entry pointing to the cluster.
> ====
> +
> [source,bash]
> ----
> # Create an ingress with a certificate
> CLUSTER_ISSUER=letsencrypt-staging
> INGRESS_HOST=cm-test.<CLUSTER_DOMAIN>
> kubectl create ingress cm-test --class=default \
>   --rule="${INGRESS_HOST}/=svc:https,tls=cm-test-ingress-cert" \
>   --annotation="cert-manager.io/cluster-issuer=${CLUSTER_ISSUER}"
> 
> # Wait for the certificate to become ready
> kubectl get certificate cm-test-ingress-cert --watch
> 
> # Delete the newly created ingress
> kubectl delete ingress cm-test
> ----

Reviewer: ```suggestion
[NOTE]
====
The cluster must be publicly reachable and have a valid DNS entry pointing to the ingress controller.
If you like to test the renewal on an existing ingress, be sure to backup the TLS Secret first.
====
+
[source,bash]
----
# Create an ingress with a certificate
CLUSTER_ISSUER=letsencrypt-staging
INGRESS_HOST=<your-DNS-entry>
kubectl create ingress cm-test --class=default \
  --rule="${INGRESS_HOST}/=svc:https,tls=cm-test-ingress-cert" \
  --annotation="cert-manager.io/cluster-issuer=${CLUSTER_ISSUER}"

# Wait for the certificate to become ready
kubectl get certificate cm-test-ingress-cert --watch

# Delete the newly created ingress
kubectl delete ingress cm-test
----
```
